File: lcode/ex01/minDepth.py
#!/usr/bin/env python
#-*- coding:utf-8 -*-

'''
Filename: minDepth
Created Date: 2019/10/9 9:59
Author: Shine

Copyright (c) 2019 41sh.cn
'''

# Definition for a binary tree node.
class TreeNode(object):
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Solution(object):
    def minDepth(self, root):
        """
        :type root: TreeNode
        :rtype: int
        """
        if root is None:
            return 0
        return self._find(root, 1)
    
    def _find(self, root, level):
        if root.left is None and root.right is None:
            return level
        elif root.left is None and root.right != None:
            return self._find(root.right, level)
        elif root.left != None and root.right == None:
            return self._find(root.left, level+1)
        else:
            return min(self._find(root.left, level+1), self._find(root.right, level+1))

File: lcode/ex34/mindepth.go
// Copyright 2019 readailib.com. All rights reserved.
// Use of this source code is governed by a BSD-style
// license that can be found in the LICENSE file.
package ex34

/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func minDepth(root *TreeNode) int {
	if root == nil {
		return 0
	}
	return walk(root, 1)
}

func walk(root *TreeNode, level int) int {
	if root.Left == nil && root.Right == nil {
		return level
	} else if root.Left == nil && root.Right != nil {
		return walk(root.Right, level+1)
	} else if root.Left != nil && root.Right == nil {
		return walk(root.Left, level+1)
	} else {
		left := walk(root.Left, level +1)
		right := walk(root.Right, level+1)
		return min(left, right)
	}
}

func min(a, b int) int {
	if a > b {
		return b
	}
	return a
}
